src/app/qgsmaptoolscalefeature.cpp

```diff
@@ -0,0 +1,479 @@
+  {
+    QLabel *lbl = new QLabel( label, this );
+    lbl->setAlignment( Qt::AlignRight | Qt::AlignCenter );
+    mLayout->addWidget( lbl );
+  }
+
+  mScaleSpinBox = new QgsDoubleSpinBox( this );
+  mScaleSpinBox->setSingleStep( 0.5 );
+  mScaleSpinBox->setMinimum( 0 );
+  mScaleSpinBox->setValue( 1 );
+  mScaleSpinBox->setShowClearButton( false );
+  mScaleSpinBox->setSizePolicy( QSizePolicy::MinimumExpanding, QSizePolicy::Preferred );
+  mLayout->addWidget( mScaleSpinBox );
+
+  mMagnetSpinBox = new QgsSpinBox( this );
+  mMagnetSpinBox->setMinimum( 0 );
+  mMagnetSpinBox->setMaximum( 180 );
+  mMagnetSpinBox->setPrefix( tr( "Snap to " ) );
+  mMagnetSpinBox->setSuffix( tr( "°" ) );
+  mMagnetSpinBox->setSingleStep( 5 );
+  mMagnetSpinBox->setValue( 0 );
+  mMagnetSpinBox->setClearValue( 0, tr( "No snapping" ) );
+  //mMagnetSpinBox->setSizePolicy( QSizePolicy::MinimumExpanding, QSizePolicy::Preferred );
+  mMagnetSpinBox->setMinimumWidth( 120 );
+  mLayout->addWidget( mMagnetSpinBox );
+
+  // connect signals
+  mScaleSpinBox->installEventFilter( this );
+  connect( mScaleSpinBox, static_cast < void ( QgsDoubleSpinBox::* )( double ) > ( &QgsDoubleSpinBox::valueChanged ), this, &QgsScaleMagnetWidget::scaleSpinBoxValueChanged );
+
+  // config focus
+  setFocusProxy( mScaleSpinBox );
+}
+
+void QgsScaleMagnetWidget::setScale( double scale )
+{
+  const int m = magnet();
+  if ( m )
+  {
+    mScaleSpinBox->setValue( std::round( scale / m ) * m );
+  }
+  else
+  {
+    mScaleSpinBox->setValue( scale );
+  }
+}
+
+double QgsScaleMagnetWidget::scale() const
+{
+  return mScaleSpinBox->value();
+}
+
+void QgsScaleMagnetWidget::setMagnet( int magnet )
+{
+  mMagnetSpinBox->setValue( magnet );
+}
+
+int QgsScaleMagnetWidget::magnet() const
+{
+  return mMagnetSpinBox->value();
+}
+
+
+bool QgsScaleMagnetWidget::eventFilter( QObject *obj, QEvent *ev )
+{
+  if ( obj == mScaleSpinBox && ev->type() == QEvent::KeyPress )
+  {
+    QKeyEvent *event = static_cast<QKeyEvent *>( ev );
+    if ( event->key() == Qt::Key_Escape )
+    {
+      emit scaleEditingCanceled();
+      return true;
+    }
+    if ( event->key() == Qt::Key_Enter || event->key() == Qt::Key_Return )
+    {
+      emit scaleEditingFinished( scale() );
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void QgsScaleMagnetWidget::scaleSpinBoxValueChanged( double scale )
+{
+  emit scaleChanged( scale );
+}
+
+QgsMapToolScaleFeature::QgsMapToolScaleFeature( QgsMapCanvas *canvas )
+  : QgsMapToolEdit( canvas )
+  , mScaling( 0 )
+  , mScalingActive( false )
+{
+}
+
+QgsMapToolScaleFeature::~QgsMapToolScaleFeature()
+{
+  deleteScalingWidget();
+  mAnchorPoint.reset();
+  deleteRubberband();
+}
+
+void QgsMapToolScaleFeature::canvasMoveEvent( QgsMapMouseEvent *e )
+{
+  if ( mScalingActive )
+  {
+    const double distance = mFeatureCenter.distance( toLayerCoordinates( mLayer, e->mapPoint() ) );
+    double scale =  distance / mBaseDistance; // min 0 or no limit?
```

Comment:
Better handle the situation where mBaseDistance == 0

Reply:
Raise an error or very small offset?

Reply:
went for return only.